src/bindings/python/flux/job/directives.py

```diff
@@ -0,0 +1,267 @@
+###############################################################
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+###############################################################
+import fileinput
+import re
+import shlex
+
+
+class Directive:
+    """
+    This class represents a single job submission directive processed
+    from an input file or batch script. Input values to the constructor
+    should have the sentinel stripped, and are then split using Python's
+    shlex module to provide familiary UNIX shell syntax and quoting
+    for argument. The first argument after shell lexing determines the
+    Directive type or "action".
+
+    At this time the only supported actions are "SETARGS", which indicates
+    an args list to pass through to the submission utility, and NOOP,
+    which is an empty directive.
+
+    If lineno is provided, it is used in error messages to indicate to
+    the user more detail about the failing line (e.g. shell parsing error).
+
+    Args:
+        value (str): A preprocessed directive with sentinel removed
+        lineno (int, optional): The source line number of the current
+            directive
+
+    Attributes:
+        lineno (int): line number associated with directive
+        args (list): list of directive arguments
+        action (str): the directive type or instruction to the submission
+            utility.
+
+    """
+
+    def __init__(self, value, lineno=-1):
+        self.lineno = lineno
+
+        if "\n" not in value:
+            #
+            # split non-multiline values as POSIX shell, removing comments.
+            # We specify posix and punctuation_chars to get more predictable
+            # quoting and avoid splitting on cmdline args like --foo.
+            lexer = shlex.shlex(value, posix=True, punctuation_chars=True)
+            #
+            # Add single-quote to escapedquotes. This is necessary to avoid
+            # unclosed quote ValueError due to escaped single quote. (The
+            # default in # Posix mode is to escape only '"')
+            lexer.escapedquotes = "\"'"
+            try:
+                self.args = list(lexer)
+            except ValueError as exc:
+                raise ValueError(f"line {lineno}: {value}: {exc}") from None
+        else:
+            self.args = [value]
```

Comment:
I'll add a comment. A multiline is a string literal, so does not get any further processing by `shlex`.

Reply:
Though now that I think about it, this may not be necessary. Multiline literals are passed to `shlex.quote()` before coming here, so I think the newlines should already be escaped. Let me test this.

Reply:
Actually, nevermind. Multilines are appended as literals at the moment, so are not quoted (only triple quotes on a single line are quoted). It will be clearer and much less error prone to add a `disable_lexer` parameter or something to the `Directive` constructor that what is currently done, so let me make that change.